### PR TITLE
Added network and application to accessories

### PR DESCRIPTION
**Pull request by agunn**
Hi Anthony

I've added the changes to allow for the network and application to be passed through with the id (I didn't remove them from the main client to minimise the code changes that would show up). Not sure if this is of any use to you. This is also the first time I've ever used GitHub so apologies if I have not done this correctly. Again, thank you for this code. I am having a fun time using it.

Next on my 'wouldn't it be nice to have list' are blind controls and irrigation....not sure if there is a way to 'Up' and 'Down' or 'Open' and 'Close' with Siri or to instruct to 'turn garden irrigation on for 10 minutes" but will have fun trying things out.

Very much look forward to seeing this progress.

All the best, Andrew

### CODE REVIEW

**Comment by anthonywebb** [comment]
Hi Andrew,

Thanks for this!  So, with this pull request, the change in the config is such that each accessory in the array will need both an application and network, correct?  Perhaps we need a fallback to use the ones defined globally if nothing is specified in the accessory?

Have you tested your changes yet?

Thanks!

File: accessories/accessory.js [hunk 1]
```diff
@@ -24,6 +24,18 @@ function CBusAccessory(platform, accessoryData)
         process.exit(0);
     }
 
+    /* We got a network? */
+    if (typeof(accessoryData.network) != "string") {
+        this.log.error("One of your accessories is missing the \"name\" field, which is required. ABORTING.");
```
Comment (anthonywebb) [inline accessories/accessory.js]: I assume you meant to tell them they are missing the "network" field?

**Comment by agunn** [comment]
Yes - Apologies!

On 19 Nov. 2016, at 10:44 am, Anthony Webb <notifications@github.com> wrote:

@anthonywebb commented on this pull request.

In accessories/accessory.js <https://github.com/anthonywebb/homebridge-cbus/pull/1#pullrequestreview-9328191>:

> @@ -24,6 +24,18 @@ function CBusAccessory(platform, accessoryData)
         process.exit(0);
     }
 
+    /* We got a network? */
+    if (typeof(accessoryData.network) != "string") {
+        this.log.error("One of your accessories is missing the \"name\" field, which is required. ABORTING.");
I assume you meant to tell them they are missing the "network" field?

**Comment by agunn** [comment]
Yes - Apologies!

On 19 Nov. 2016, at 10:44 am, Anthony Webb <notifications@github.com> wrote:

@anthonywebb commented on this pull request.

In accessories/accessory.js <https://github.com/anthonywebb/homebridge-cbus/pull/1#pullrequestreview-9328179>:

> @@ -24,6 +24,18 @@ function CBusAccessory(platform, accessoryData)
         process.exit(0);
     }
 
+    /* We got a network? */
+    if (typeof(accessoryData.network) != "string") {
+        this.log.error("One of your accessories is missing the \"name\" field, which is required. ABORTING.");
+        process.exit(0);
+    }
+		
+    /* We got a application? */
+    if (typeof(accessoryData.application) != "string") {
+        this.log.error("One of your accessories is missing the \"name\" field, which is required. ABORTING.");
I assume you meant to tell them they are missing the "application" field?

**Comment by agunn** [comment]
Agreed - I was reluctant to change your code too much - wanted to check would work with my minimal changes…

I think it would be a good idea to have a fallback


On 19 Nov. 2016, at 10:47 am, Anthony Webb <notifications@github.com> wrote:

@anthonywebb commented on this pull request.

In cgate-client.js <https://github.com/anthonywebb/homebridge-cbus/pull/1#pullrequestreview-9328455>:

> @@ -204,25 +204,25 @@ CBusClient.prototype._socketReceivedMessageEvent = function(message, type)
 //  Private API
 //==========================================================================================
 
-CBusClient.prototype._buildCommandString = function(device,command,level,delay) {
+CBusClient.prototype._buildCommandString = function(network, application, device,command,level,delay) {
     var message = '';
 
Maybe here we should listen for null network/application and use this.clientNetwork/this.clientApplication as a fallback so it isnt required to enter a network/application for every single accessory?

**Comment by agunn** [comment]
Hi Anthony - yes that’s the intention: being able to use any network and application number. The fallback would be a good idea.

I use HomeGate for control of my system running on a 27” touchscreen. 

The install has166 units -  64 units on network 254 and 82 on network 254 with around 200 GAs on the lighting app, 20+ on the Irrigation app  and 20 on the Heating app. There’s other stuff on Enable Control and Triggers so a fair bit!

I have implemented the network and application changes and they are working well. I can turn on and off lights on either network using Siri or the Home app, individually or by ‘room'. 

I would love to get irrigation and blinds going. I have 32 blind controllers and a bunch of low voltage relays for the irrigation. I tried the blinds using the WindowCovering service but am having some trouble with feedback to show the state - I can trigger the open and close but it doesn’t then show as Open or Closed in the Home app….

Really enjoying implementing this.

Thank you!


On 19 Nov. 2016, at 10:43 am, Anthony Webb <notifications@github.com> wrote:

Hi Andrew,

Thanks for this! So, with this pull request, the change in the config is such that each accessory in the array will need both an application and network, correct? Perhaps we need a fallback to use the ones defined globally if nothing is specified in the accessory?

Have you tested your changes yet?

Thanks!

**Comment by agunn** [comment]
And sorry - I don’t have a fan controller!

On 19 Nov. 2016, at 11:22 am, Andrew Gunn <a.gunn@optusnet.com.au> wrote:

Hi Anthony - yes that’s the intention: being able to use any network and application number. The fallback would be a good idea.

I use HomeGate for control of my system running on a 27” touchscreen. 

The install has166 units -  64 units on network 254 and 82 on network 254 with around 200 GAs on the lighting app, 20+ on the Irrigation app  and 20 on the Heating app. There’s other stuff on Enable Control and Triggers so a fair bit!

I have implemented the network and application changes and they are working well. I can turn on and off lights on either network using Siri or the Home app, individually or by ‘room'. 

I would love to get irrigation and blinds going. I have 32 blind controllers and a bunch of low voltage relays for the irrigation. I tried the blinds using the WindowCovering service but am having some trouble with feedback to show the state - I can trigger the open and close but it doesn’t then show as Open or Closed in the Home app….

Really enjoying implementing this.

Thank you!


On 19 Nov. 2016, at 10:43 am, Anthony Webb <notifications@github.com <mailto:notifications@github.com>> wrote:

Hi Andrew,

Thanks for this! So, with this pull request, the change in the config is such that each accessory in the array will need both an application and network, correct? Perhaps we need a fallback to use the ones defined globally if nothing is specified in the accessory?

Have you tested your changes yet?

Thanks!

**Comment by anthonywebb** [comment]
Boy Andrew, that sounds like quite a network! 

As for the WindowCovering service and not getting the updates, I had similar issues with lights too, but I assume if you are using my latest code you have a blueprint of how to get around the pitfalls I was seeing.  If you would like me to take a glance at your code I'm happy to see if anything jumps out at me.

**Comment by agunn** [comment]
Thanks Anthony I might do that after I play around with it a bit today.

I have your latest code. If you can suggest any decent places for docs I’d be grateful. There doesn’t seem to be much…


 
On 19 Nov. 2016, at 11:29 am, Anthony Webb <notifications@github.com> wrote:

Boy Andrew, that sounds like quite a network!

As for the WindowCovering service and not getting the updates, I had similar issues with lights too, but I assume if you are using my latest code you have a blueprint of how to get around the pitfalls I was seeing. If you would like me to take a glance at your code I'm happy to see if anything jumps out at me.

**Comment by agunn** [comment]
I’ve got the HomeKit docs from Apple Dev and the HomeKitTypes.js stuff but if there’s something more explicit it would be great...

On 19 Nov. 2016, at 12:09 pm, Andrew Gunn <a.gunn@optusnet.com.au> wrote:

Thanks Anthony I might do that after I play around with it a bit today.

I have your latest code. If you can suggest any decent places for docs I’d be grateful. There doesn’t seem to be much…


 
On 19 Nov. 2016, at 11:29 am, Anthony Webb <notifications@github.com <mailto:notifications@github.com>> wrote:

Boy Andrew, that sounds like quite a network!

As for the WindowCovering service and not getting the updates, I had similar issues with lights too, but I assume if you are using my latest code you have a blueprint of how to get around the pitfalls I was seeing. If you would like me to take a glance at your code I'm happy to see if anything jumps out at me.

**Comment by agunn** [comment]
Hi Anthony - 

Sorry not to get back to you sooner I have been away. Although I have no duplicate GAs across the networks in the lighting group there are overlaps between lighting and other apps (heating, irrigation) so getting the following:

/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/Accessory.js:222
      throw new Error("Cannot add a bridged Accessory with the same UUID as another bridged Accessory: " + existing.UUID);
 
Was wondering if multiplying the app # by (say) 1000 and adding the GA to create a pseudo UUID would solve this (rather than just using the C-Bus GA on it’s own). So for instance GA 225 in lighting would be GA 56225 and (say) the same group in the Heating app would be 136225.

Any thoughts? Would obviously have to be only the code that registers the accessory and ensure that the proper app and GA are sent to C-Bus.

Thanks 

Andrew

 

Have come across interesting issue using multiple networks/apps.


On 19 Nov. 2016, at 11:29 am, Anthony Webb <notifications@github.com> wrote:

Boy Andrew, that sounds like quite a network!

As for the WindowCovering service and not getting the updates, I had similar issues with lights too, but I assume if you are using my latest code you have a blueprint of how to get around the pitfalls I was seeing. If you would like me to take a glance at your code I'm happy to see if anything jumps out at me.

**Comment by gbrooker** [comment]
Anthony, Andrew, 
please take a look at my proposed alternative, which allows the network and application ID's to be optionally specified for any accessory. If they are not specified they default to the project network and default application. This also encodes the network and application into the uuid to ensure there are no clashes. Note that you will probably have to clear the cache by deleting files in ~/.homebridge/persist/

Cheers
Guy

**Comment by anthonywebb** [comment]
Thanks @gbrooker, I merged this.  @agunn I think you will like the merge, if there are things missing you would like to see, lets discuss.

**Comment by agunn** [comment]
That’s great thanks. Will take a look.


On 12 Jan 2017, at 4:54 am, Anthony Webb <notifications@github.com> wrote:

Thanks @gbrooker <https://github.com/gbrooker>, I merged this. @agunn <https://github.com/agunn> I think you will like the merge, if there are things missing you would like to see, lets discuss.

**Comment by simonhac** [comment]
@anthonywebb i happened across this thread. as the implementation currently sits, accessories get a different UUID for every invocation of `homebridge-cbus`. do you think this is a problem? might things behave a little differently if the UUIDs were constant across invocations?

**Comment by anthonywebb** [comment]
If we could come up with a consistent uuid that would be ideal.

**Comment by simonhac** [comment]
@anthonywebb do you know if it'd make a difference? 

why not use the moduleId (which i think i renamed to hash)?

**Comment by anthonywebb** [comment]
That would make sense to me. Let's try.

**Pull request closed without merging**